public/app/plugins/panel/geomap/components/BaseMap.tsx

```diff
@@ -0,0 +1,199 @@
+import React, { Component } from 'react';
+import { geomapLayerRegistry } from '../layers/registry';
+import { Map as GeoMap, View } from 'ol';
+import Attribution from 'ol/control/Attribution';
+import Zoom from 'ol/control/Zoom';
+import ScaleLine from 'ol/control/ScaleLine';
+import BaseLayer from 'ol/layer/Base';
+
+import { PanelData, MapLayerHandler, MapLayerConfig } from '@grafana/data';
+import { config } from '@grafana/runtime';
+
+interface BaseMapProps {
+  width: number;
+  height: number;
+  options: GeomapPanelOptions;
+  data: PanelData;
+}
+
+import 'ol/ol.css';
+
+import { ControlsOptions, GeomapPanelOptions } from '../types';
+import { defaultGrafanaThemedMap } from '../layers/basemaps/theme';
+import { InfoControl } from './InfoControl';
+
+interface MapLayerState {
+  config: MapLayerConfig;
+  handler: MapLayerHandler;
+  layer: BaseLayer; // used to add|remove
+}
+
+export class BaseMap extends Component<BaseMapProps> {
+  map: GeoMap;
+
+  basemap: BaseLayer;
+  layers: MapLayerState[] = [];
+
+  constructor(props: BaseMapProps) {
+    super(props);
+  }
+
+  componentDidUpdate(oldProps: BaseMapProps) {
+    if (!this.map) {
+      console.log('SKIPPING????');
+      return; // not yet initalized
+    }
+
+    // Check for resize
+    if (this.props.height !== oldProps.height || this.props.width !== oldProps.width) {
+      this.map.updateSize();
+    }
+
+    // External configuraiton changed
+    if (this.props.options !== oldProps.options) {
+      this.optionsChanged(oldProps.options);
+    }
+
+    // External data changed
+    if (this.props.data !== oldProps.data) {
+      this.dataChanged(this.props.data);
+    }
+  }
+
+  /**
+   * Called when the panel options change
+   */
+  optionsChanged(oldOptions: GeomapPanelOptions) {
+    const { options } = this.props;
+    console.log('options changed!', options);
+
+    // NOTE: the panel options editor currently mutates nested objects
+    // this means we can not easily detect changes to layer settings :(
+    // for now, lets just re-init everything whenever something changes
```

Comment:
@ryantxu did some investigation on this today, and TBH I'm super confused. Have no idea why the mutation would happen as we do entire panel options update on a single option change (https://github.com/grafana/grafana/blob/f0cac149dacfff9be191161979e083daa2f91821/public/app/features/dashboard/state/PanelModel.ts#L247)

Even deep cloning of the options doesn't help. At the same time, debuging previous vs current props of a component definitely shows that these objects are mutated (see this branch https://github.com/grafana/grafana/tree/debug-options - use the Debug panel to see the prev props of a panel are actually same as current regardless of the change in the UI).


Have not come to any conclussions yet, unfortunately.

Reply:
Wait, shit, i now see lodash set used here: https://github.com/grafana/grafana/blob/2a624993d27dbc222c461f4ab49a326b45bae50a/public/app/features/dashboard/components/PanelEditor/getVizualizationOptions.tsx#L58

Reply:
with deep cloning options passed to the options editor it seems to work, but i think this may have negative performance impact: https://github.com/grafana/grafana/commit/60a5dce36c97e1a7329ff1eb51829dbfe8fd19c1

Will think more about way to replace _.set with something else (dunno what yet :P ). But the lodash suggested way of doing immutable set (https://github.com/lodash/lodash/issues/1696#issuecomment-328335502) does not work well in our case too, as the parent object references will still be the same :/

Reply:
I merged in: #36441  and things work great!